src/PrestaShopBundle/Controller/Admin/Improve/Design/CmsPageController.php

```diff
@@ -256,6 +256,11 @@ public function editAction(Request $request, $cmsPageId)
                     'cmsPageId' => $cmsPageId,
                 ]),
             ]);
+
+            /** @var EditableCmsPage $editableCmsPage */
+            $editableCmsPage = $this->getQueryBus()->handle(new GetCmsPageForEditing($cmsPageId));
+            $previewUrl = $editableCmsPage->getPreviewUrl();
```

Comment:
If you move this statement, then we're going to present the `$previewUrl` *before* the update. Looks like an issue 🤔. If the merchant modifies the CMS Page and what should be previewed, after update he will still see the outdated preview.

Reply:
:thinking: I doubt it's an issue. In form you cannot change anything that would impact the `$previewUrl` itself. Url contains id_cms and id_lang (those aren't configurable in edit action). In addition - if `$previewUrl` is defined after handler it might cause troubles in case handler throws an exception (the redirect page cannot be rendered as the `$previewUrl` is not yet defined). Does it make sense, or am I missing something?

Reply:
> 🤔 I doubt it's an issue. In form you cannot change anything that would impact the `$previewUrl` itself. Url contains id_cms and id_lang (those aren't configurable in edit action).

That's a good point 👍 

However
>  In addition - if `$previewUrl` is defined after handler it might cause troubles in case handler throws an exception (the redirect page cannot be rendered as the `$previewUrl` is not yet defined). Does it make sense, or am I missing something?

You have raised an important issue 😉 if `$previewUrl` is used to render the page, then it cannot be in the `try/catch` block as there is a possibility we get out of the `try` without assigning the variable.

We have 2 solutions to handle it:
- how is `$previewUrl` used in twig ? can we pass a default value like an empty string ?
- if not, then we must check, just before the render, `if ($previewUrl === null)` and handle the usecase*

*By handle the usecase, I mean:
- if the previewUrl is a critical item of the page, we throw an exception
- if the previewUrl is a mandatory item of the page, we display a red error message in twig 
- if the previewUrl is n optional item of the page, we display a orange warning message in twig

Reply:
Isn't it enough to handle it here: https://github.com/PrestaShop/PrestaShop/blob/03aa43cfb7745f1c6e2a8df86e666e4cdb83e499/src/PrestaShopBundle/Controller/Admin/Improve/Design/CmsPageController.php#L288 ?
PreviewUrl is a property of EditableCmsPage, so it could be missing only if EditableCmsPage is not found. In that case it's redirecting to index with "The object cannot be loaded (or found)" error

Reply:
No it could be missing if anything else goes wrong in `$this->getQueryBus()->handle(new GetCmsPageForEditing($cmsPageId));` 😄

Reply:
https://github.com/PrestaShop/PrestaShop/pull/14168/commits/a2759bfbf639cbf1d1621f0933b1330910d76355
wdyt ? Ensures the the $form and $editableCmsPage are defined if reaches return (code looks terrible with 2 tries though). Same case in CategoryController edit action.